File: 2018summer/07.12/solutions_2015/bountyhii/submissions/accepted/tobi.java
import java.util.*;

public class tobi {
	static ArrayList<Integer> adj[] = new ArrayList[2000];
	static int pred[] = new int[2000], N, K, to;
	static boolean mark[] = new boolean[2000];

	static boolean match(int n) {
		if (n == -1) return true;
		for (int to : adj[n]) {
			if (mark[to]) continue;
			mark[to] = true;
			if (match(pred[to])) {
				pred[to] = n;
				return true;
			}
		}
		return false;
	}

	static int maxmatch() {
		Arrays.fill(pred, -1);
		int ret = 0;
		for (int i = 0; i < N; i++) {
			Arrays.fill(mark, false);
			if (match(i)) ret++;
		}
		return ret;
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		N = sc.nextInt();
		for (int i = 0; i < N; i++) {
			K = sc.nextInt();
			adj[i] = new ArrayList<>();
			for (int j = 0; j < K; j++) adj[i].add(sc.nextInt());
		}
		System.out.println(N-maxmatch());
	}
}

File: 2018summer/07.12/solutions_2015/bountyhii/submissions/accepted/tobi.cpp
#include <iostream>
#include <vector>

using namespace std;
#define FOR(i,a,b) for (int i = (a); i < (b); i++)

vector<int> adj[2000];
int pred[2000], N, K, to;
bool mark[2000];

bool match(int n) {
	if (n == -1) return true;
	for (int to : adj[n]) {
		if (mark[to]) continue;
		mark[to] = true;
		if (match(pred[to])) {
			pred[to] = n;
			return true;
		}
	}
	return false;
}

int maxmatch() {
	fill_n(pred, 2000, -1);
	int ret = 0;
	FOR(i, 0, N) {
		fill_n(mark, 2000, false);
		if (match(i)) ret++;
	}
	return ret;
}

int main() {
	cin >> N;
	FOR(i, 0, N) {
		cin >> K;
		FOR(j, 0, K) {
			cin >> to;
			adj[i].push_back(to);
		}
	}
	cout << N-maxmatch() << endl;

	return 0;
}
